akka-docs/src/test/java/jdocs/stream/operators/flow/StatefulMap.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package jdocs.stream.operators.flow;
+
+import akka.actor.typed.ActorSystem;
+import akka.japi.Pair;
+import akka.stream.javadsl.Flow;
+import akka.stream.javadsl.Source;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class StatefulMap {
+    static final ActorSystem<?> system = null;
+
+    public void bufferUntilChanged() {
+        // #bufferUntilChanged
+        Source.from(Arrays.asList("A", "B", "B", "C", "C", "C", "D"))
+            .statefulMap(() -> (List<String>) new LinkedList<String>(),
+                (list, element) -> {
+                    if (list.isEmpty()) {
+                        list.add(element);
+                        return Pair.create(list, Collections.<String>emptyList());
+                    } else if (list.get(0).equals(element)) {
+                        list.add(element);
+                        return Pair.create(list, Collections.<String>emptyList());
+                    } else {
+                        final List<String> newBuffer = new LinkedList<>();
+                        newBuffer.add(element);
+                        return Pair.create(newBuffer, list);
+                    }
+                }, Optional::ofNullable)
+            .runForeach(System.out::println, system);
+        //prints
+        //[A]
+        //[B, B]
+        //[C, C, C]
+        //[D]
+        //#bufferUntilChanged
+    }
+
+    public void distinctUntilChanged() {
+        //#distinctUntilChanged
+        Source.from(Arrays.asList("A", "B", "B", "C", "C", "C", "D"))
+            .statefulMap(Optional::<String>empty,
+                (lastElement, element) -> {
+                    if (lastElement.isPresent() && lastElement.get().equals(element)) {
+                        return Pair.create(lastElement, Optional.<String>empty());
+                    } else {
+                        return Pair.create(Optional.of(element), Optional.of(element));
+                    }
+                }, Optional::ofNullable)
+            .via(Flow.flattenOptional())
+            .runForeach(System.out::println, system);
+        //prints
+        //A
+        //B
+        //C
+        //D
+        //#distinctUntilChanged
+    }
+
+    public void usingResource() {
+        // #usingResource
+        Source.from(Arrays.asList("A", "B", "C", "D"))
+            .statefulMap(() -> {
+                final AtomicInteger resource = new AtomicInteger(0);
+                resource.set(100);
+                return resource;
+            }, (resource, element) -> {
+                //mock resource usage
+                final String output = "element " + element + " use resource: " + resource.getAndDecrement();
+                return Pair.create(resource, output);
+            }, r -> {
+                r.set(0);
+                return Optional.empty();
+            }).runForeach(System.out::println, system);
```

Comment:
use as resource

Reply:
Actually, I wonder if we shouldn't skip this sample and add a separate operator `withResource`/`mapWithResource` that delegates to `statefulMap` but correctly sets the safe blocking io dispatcher.

If we keep it here I think this one would be better with an actual pretend resource like in the `jdocs.stream.operators.source.UnfoldResource` sample.

And since it is showing a resource it will quite likely be blocking, so sample should also run it on the blocking io dispatcher, or else this will be copy-pasted into user code and used in dangerous ways.

Reply:
ok, let's leave that part as another task.

Reply:
Created https://github.com/akka/akka/issues/31355 to track that